src/bin/devmgr/src/nvme_test.rs

```diff
@@ -57,8 +78,273 @@ impl RequestDriver for NvmeQueue {
+    fn get_prp_list_or_buffer(&mut self) -> Option<nvme::ds::cmd::PrpListOrBuffer> {
+        todo!()
+    }
+
+    fn get_dptr(&mut self, _sgl_allowed: bool) -> Option<nvme::ds::queue::subentry::Dptr> {
+        let pin = self.0.pin().unwrap();
+        Some(nvme::ds::queue::subentry::Dptr::Prp(
+            pin[0].addr().into(),
+            0,
+        ))
+    }
+}
+
+struct NvmeRequester {
+    subq: Mutex<SubmissionQueue>,
+    comq: Mutex<CompletionQueue>,
+    sub_bell: *const VolatileCell<u32>,
+    com_bell: *const VolatileCell<u32>,
+}
+
+unsafe impl Send for NvmeRequester {}
+unsafe impl Sync for NvmeRequester {}
+
+impl NvmeRequester {
+    fn check_completions(&self) -> Vec<ResponseInfo<CommonCompletion>> {
+        let mut comq = self.comq.lock().unwrap();
+        let mut resps = Vec::new();
+        let mut new_head = None;
+        let mut new_bell = None;
+        while let Some((bell, resp)) = comq.get_completion::<CommonCompletion>() {
+            let id: u16 = resp.command_id().into();
+            //println!("got completion for {} {} {}", resp.new_sq_head(), bell, id);
+            resps.push(ResponseInfo::new(resp, id as u64, false));
+            new_head = Some(resp.new_sq_head());
+            new_bell = Some(bell);
+        }
+
+        if let Some(head) = new_head {
+            self.subq.lock().unwrap().update_head(head);
+        }
+
+        if let Some(bell) = new_bell {
+            unsafe { self.com_bell.as_ref().unwrap().set(bell as u32) }
+        }
+
+        resps
+    }
+}
+
+#[async_trait::async_trait]
+impl RequestDriver for NvmeRequester {
+    type Request = CommonCommand;
+
+    type Response = CommonCompletion;
+
+    type SubmitError = ();
+
+    async fn submit(
+        &self,
+        reqs: &mut [SubmitRequest<Self::Request>],
+    ) -> Result<(), Self::SubmitError> {
+        let mut sq = self.subq.lock().unwrap();
+        let mut tail = None;
+        for sr in reqs.iter_mut() {
+            let cid = (sr.id() as u16).into();
+            sr.data_mut().set_cid(cid);
+            tail = sq.submit(sr.data());
+            assert!(tail.is_some());
+        }
+        if let Some(tail) = tail {
+            unsafe {
+                self.sub_bell.as_ref().unwrap().set(tail as u32);
+            }
+        }
+        Ok(())
+    }
+
+    fn flush(&self) {}
+
+    const NUM_IDS: usize = 32;
+}
+
+fn init_controller(ctrl: &mut Arc<NvmeController>) {
+    let bar = ctrl.device_ctrl.device().get_mmio(1).unwrap();
+    let reg =
+        unsafe { bar.get_mmio_offset::<nvme::ds::controller::properties::ControllerProperties>(0) };
+
+    let config = ControllerConfig::new();
+    reg.configuration.set(config);
+
+    while reg.status.get().ready() {
+        core::hint::spin_loop();
+    }
+    println!("version: {} {}", reg.version_maj(), reg.version_min());
+
+    let aqa = nvme::ds::controller::properties::aqa::AdminQueueAttributes::new()
+        .with_completion_queue_size(32 - 1)
```

Comment:
Is this 32 supposed to be NUM_IDS?

Reply:
Yes, this is one thing I will fix when I rewrite this driver, but I wanted to keep the driver rewrite in a separate PR.